# **Pre-lecture HW**:

## **1.**

In [ ]:
import numpy as np
import pandas as pd

In [ ]:
# URL of the dataset
url = 'https://raw.githubusercontent.com/tmsllab/datasets/main/Churn_Modelling.csv'

# Load the dataset into a DataFrame
df = pd.read_csv(url)

df

RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
0             1    15634602   Hargrave          619    France  Female   42   
1             2    15647311       Hill          608     Spain  Female   41   
2             3    15619304       Onio          502    France  Female   42   
3             4    15701354       Boni          699    France  Female   39   
4             5    15737888   Mitchell          850     Spain  Female   43   
...         ...         ...        ...          ...       ...     ...  ...   
9995       9996    15606229   Obijiaku          771    France    Male   39   
9996       9997    15569892  Johnstone          516    France    Male   35   
9997       9998    15584532        Liu          709    France  Female   36   
9998       9999    15682355  Sabbatini          772   Germany    Male   42   
9999      10000    15628319     Walker          792    France  Female   28   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0          2       0.00              1          1               1   
1          1   83807.86              1          0               1   
2          8  159660.80              3          1               0   
3          1       0.00              2          0               0   
4          2  125510.82              1          1               1   
...      ...        ...            ...        ...             ...   
9995       5       0.00              2          1               0   
9996      10   57369.61              1          1               1   
9997       7       0.00              1          0               1   
9998       3   75075.31              2          1               0   
9999       4  130142.79              1          1               0   

      EstimatedSalary  Exited  
0           101348.88       1  
1           112542.58       0  
2           113931.57       1  
3            93826.63       0  
4            79084.10       0  
...               ...     ...  
9995         96270.64       0  
9996        101699.77       0  
9997         42085.58       1  
9998         92888.52       1  
9999         38190.78       0  

[10000 rows x 14 columns]

In [ ]:
# Check for missing values
missing_values = df.isnull().sum()

# Display the result
missing_values

RowNumber          0
CustomerId         0
Surname            0
CreditScore        0
Geography          0
Gender             0
Age                0
Tenure             0
Balance            0
NumOfProducts      0
HasCrCard          0
IsActiveMember     0
EstimatedSalary    0
Exited             0
dtype: int64

From here, we can conclude that there is no missing values in the dataset above.

## **2.**

In [ ]:
# Get the number of columns and rows
num_columns = df.shape[1]
num_rows = df.shape[0]

print(f'Number of columns: {num_columns}')
print(f'Number of rows: {num_rows}')


Number of columns: 14
Number of rows: 10000


Definition of "observations": Observations correspond to the individual data points or records that have been collected. In a dataset, each row represents a single observation.

Definition of "variables": Variables represent the variety of properties being measured or collected for each observation. In a dataset, variables are the columns, and they describe the different types of data points collected for each observation.

## **3.**

In [ ]:
# Get a summary of the dataset
summary = df.describe()
summary

RowNumber    CustomerId   CreditScore           Age        Tenure  \
count  10000.00000  1.000000e+04  10000.000000  10000.000000  10000.000000   
mean    5000.50000  1.569094e+07    650.528800     38.921800      5.012800   
std     2886.89568  7.193619e+04     96.653299     10.487806      2.892174   
min        1.00000  1.556570e+07    350.000000     18.000000      0.000000   
25%     2500.75000  1.562853e+07    584.000000     32.000000      3.000000   
50%     5000.50000  1.569074e+07    652.000000     37.000000      5.000000   
75%     7500.25000  1.575323e+07    718.000000     44.000000      7.000000   
max    10000.00000  1.581569e+07    850.000000     92.000000     10.000000   

             Balance  NumOfProducts    HasCrCard  IsActiveMember  \
count   10000.000000   10000.000000  10000.00000    10000.000000   
mean    76485.889288       1.530200      0.70550        0.515100   
std     62397.405202       0.581654      0.45584        0.499797   
min         0.000000       1.000000      0.00000        0.000000   
25%         0.000000       1.000000      0.00000        0.000000   
50%     97198.540000       1.000000      1.00000        1.000000   
75%    127644.240000       2.000000      1.00000        1.000000   
max    250898.090000       4.000000      1.00000        1.000000   

       EstimatedSalary        Exited  
count     10000.000000  10000.000000  
mean     100090.239881      0.203700  
std       57510.492818      0.402769  
min          11.580000      0.000000  
25%       51002.110000      0.000000  
50%      100193.915000      0.000000  
75%      149388.247500      0.000000  
max      199992.480000      1.000000

In [ ]:
summary_all = df.describe(include='all')
summary_all

RowNumber    CustomerId Surname   CreditScore Geography Gender  \
count   10000.00000  1.000000e+04   10000  10000.000000     10000  10000   
unique          NaN           NaN    2932           NaN         3      2   
top             NaN           NaN   Smith           NaN    France   Male   
freq            NaN           NaN      32           NaN      5014   5457   
mean     5000.50000  1.569094e+07     NaN    650.528800       NaN    NaN   
std      2886.89568  7.193619e+04     NaN     96.653299       NaN    NaN   
min         1.00000  1.556570e+07     NaN    350.000000       NaN    NaN   
25%      2500.75000  1.562853e+07     NaN    584.000000       NaN    NaN   
50%      5000.50000  1.569074e+07     NaN    652.000000       NaN    NaN   
75%      7500.25000  1.575323e+07     NaN    718.000000       NaN    NaN   
max     10000.00000  1.581569e+07     NaN    850.000000       NaN    NaN   

                 Age        Tenure        Balance  NumOfProducts    HasCrCard  \
count   10000.000000  10000.000000   10000.000000   10000.000000  10000.00000   
unique           NaN           NaN            NaN            NaN          NaN   
top              NaN           NaN            NaN            NaN          NaN   
freq             NaN           NaN            NaN            NaN          NaN   
mean       38.921800      5.012800   76485.889288       1.530200      0.70550   
std        10.487806      2.892174   62397.405202       0.581654      0.45584   
min        18.000000      0.000000       0.000000       1.000000      0.00000   
25%        32.000000      3.000000       0.000000       1.000000      0.00000   
50%        37.000000      5.000000   97198.540000       1.000000      1.00000   
75%        44.000000      7.000000  127644.240000       2.000000      1.00000   
max        92.000000     10.000000  250898.090000       4.000000      1.00000   

        IsActiveMember  EstimatedSalary        Exited  
count     10000.000000     10000.000000  10000.000000  
unique             NaN              NaN           NaN  
top                NaN              NaN           NaN  
freq               NaN              NaN           NaN  
mean          0.515100    100090.239881      0.203700  
std           0.499797     57510.492818      0.402769  
min           0.000000        11.580000      0.000000  
25%           0.000000     51002.110000      0.000000  
50%           1.000000    100193.915000      0.000000  
75%           1.000000    149388.247500      0.000000  
max           1.000000    199992.480000      1.000000

## **4.**

In [ ]:
url2 = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'

df2 = pd.read_csv(url2)

df2

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [ ]:
df2.shape

(891, 15)

In [ ]:
df2.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

For the difference in number of columns, it is caused by the function `.describe()` not including columns with non-numeric values on them. For the difference in the number of `count` in the column `age` not correspond with the number of rows, the `count` in `.describe()` only shows the number of non-NA values (or non-missing values), which means that there are exactly 177 missing values in the `age` column.

In [ ]:
df2.describe(include="all")

survived      pclass   sex         age       sibsp       parch  \
count   891.000000  891.000000   891  714.000000  891.000000  891.000000   
unique         NaN         NaN     2         NaN         NaN         NaN   
top            NaN         NaN  male         NaN         NaN         NaN   
freq           NaN         NaN   577         NaN         NaN         NaN   
mean      0.383838    2.308642   NaN   29.699118    0.523008    0.381594   
std       0.486592    0.836071   NaN   14.526497    1.102743    0.806057   
min       0.000000    1.000000   NaN    0.420000    0.000000    0.000000   
25%       0.000000    2.000000   NaN   20.125000    0.000000    0.000000   
50%       0.000000    3.000000   NaN   28.000000    0.000000    0.000000   
75%       1.000000    3.000000   NaN   38.000000    1.000000    0.000000   
max       1.000000    3.000000   NaN   80.000000    8.000000    6.000000   

              fare embarked  class  who adult_male deck  embark_town alive  \
count   891.000000      889    891  891        891  203          889   891   
unique         NaN        3      3    3          2    7            3     2   
top            NaN        S  Third  man       True    C  Southampton    no   
freq           NaN      644    491  537        537   59          644   549   
mean     32.204208      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
std      49.693429      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
min       0.000000      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
25%       7.910400      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
50%      14.454200      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
75%      31.000000      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
max     512.329200      NaN    NaN  NaN        NaN  NaN          NaN   NaN   

       alone  
count    891  
unique     2  
top     True  
freq     537  
mean     NaN  
std      NaN  
min      NaN  
25%      NaN  
50%      NaN  
75%      NaN  
max      NaN

## **5.**

Attribute is a property of an object that holds a specific infomation about the object itself.

Method is a function that performs a specific action to an object. This specific action may or may not use a parameter, hence the bracket at the end of it.

## ChatGPT link (for q1): https://chatgpt.com/share/70e8d655-6789-480c-b0d0-7408b183711a

## ChatGPT link (for q2-q5): https://chatgpt.com/share/4ef7defd-eb8d-451b-84c8-8aacf8e18a23 (Some of the codes I've written in q2-q5 are slightly different because there is this previous [conversation](https://chatgpt.com/share/7b6061d6-b52f-4043-9da0-60ddd2c18065), but as I want ChatGPT to re-summarize (to include the ones before the new dataset I've given to ChatGPT), my GPT-4o use limit has been reached, thus making me unable to continue the conversation.)

# **Post-lecture HW:**

## **6.**

`count`: Provides the number of non-missing values for each column.

`mean`: The average value of all non-missing values for each column.

`std`: The standard deviation of all non-missing values for each column. The higher the standard deviation, the more the data spread in a wider range

`min`: The lowest value of all non-missing values for each column.

`25%`: The value such that exactly 25% of all non-missing values are lower than the value itself, for each column.

`50%`: The median value of all non-missing values for each column. Median value is a value such that the number of values below the median equals the number of values above the median.

`75%`: The value such that exactly 75% of all non-missing values are lower than the value itself, for each column

`max`: The highest value of all non-missing values for each column.

## **7.**

1. When you want to remove rows (or columns if you use `df.dropna(axis=1)`) that contains missing values.
2. When you want to remove columns that is considered as not useful for the dataset (e.g. irrelevant columns, columns with too many missing data)
3. Applying `del df['col']` before `df.dropna()` could be important in the case that you want to remove rows and columns that are not useful (as mentioned in 2.). Since executing `del df['col']` removes columns with too many missing data, doing that before `df.dropna()` preserves some rows (or observations) that would otherwise be deleted by `df.dropna()` since they only have the missing data in the "not useful" column.

In [ ]:
df2.isna().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [ ]:
df2_dropdel = df2.dropna()
del df2_dropdel['deck'], df2_dropdel['age']
df2_dropdel.describe(include='all')

survived      pclass   sex       sibsp       parch        fare  \
count   182.000000  182.000000   182  182.000000  182.000000  182.000000   
unique         NaN         NaN     2         NaN         NaN         NaN   
top            NaN         NaN  male         NaN         NaN         NaN   
freq           NaN         NaN    94         NaN         NaN         NaN   
mean      0.675824    1.192308   NaN    0.467033    0.478022   78.919735   
std       0.469357    0.516411   NaN    0.645007    0.755869   76.490774   
min       0.000000    1.000000   NaN    0.000000    0.000000    0.000000   
25%       0.000000    1.000000   NaN    0.000000    0.000000   29.700000   
50%       1.000000    1.000000   NaN    0.000000    0.000000   57.000000   
75%       1.000000    1.000000   NaN    1.000000    1.000000   90.000000   
max       1.000000    3.000000   NaN    3.000000    4.000000  512.329200   

       embarked  class  who adult_male  embark_town alive  alone  
count       182    182  182        182          182   182    182  
unique        3      3    3          2            3     2      2  
top           S  First  man      False  Southampton   yes  False  
freq        115    157   87         95          115   123    104  
mean        NaN    NaN  NaN        NaN          NaN   NaN    NaN  
std         NaN    NaN  NaN        NaN          NaN   NaN    NaN  
min         NaN    NaN  NaN        NaN          NaN   NaN    NaN  
25%         NaN    NaN  NaN        NaN          NaN   NaN    NaN  
50%         NaN    NaN  NaN        NaN          NaN   NaN    NaN  
75%         NaN    NaN  NaN        NaN          NaN   NaN    NaN  
max         NaN    NaN  NaN        NaN          NaN   NaN    NaN

In [ ]:
df2_deldrop = df2.copy()
del df2_deldrop['deck'], df2_deldrop['age']
df2_deldrop = df2_deldrop.dropna()
df2_deldrop.describe(include='all')

survived      pclass   sex       sibsp       parch        fare  \
count   889.000000  889.000000   889  889.000000  889.000000  889.000000   
unique         NaN         NaN     2         NaN         NaN         NaN   
top            NaN         NaN  male         NaN         NaN         NaN   
freq           NaN         NaN   577         NaN         NaN         NaN   
mean      0.382452    2.311586   NaN    0.524184    0.382452   32.096681   
std       0.486260    0.834700   NaN    1.103705    0.806761   49.697504   
min       0.000000    1.000000   NaN    0.000000    0.000000    0.000000   
25%       0.000000    2.000000   NaN    0.000000    0.000000    7.895800   
50%       0.000000    3.000000   NaN    0.000000    0.000000   14.454200   
75%       1.000000    3.000000   NaN    1.000000    0.000000   31.000000   
max       1.000000    3.000000   NaN    8.000000    6.000000  512.329200   

       embarked  class  who adult_male  embark_town alive alone  
count       889    889  889        889          889   889   889  
unique        3      3    3          2            3     2     2  
top           S  Third  man       True  Southampton    no  True  
freq        644    491  537        537          644   549   535  
mean        NaN    NaN  NaN        NaN          NaN   NaN   NaN  
std         NaN    NaN  NaN        NaN          NaN   NaN   NaN  
min         NaN    NaN  NaN        NaN          NaN   NaN   NaN  
25%         NaN    NaN  NaN        NaN          NaN   NaN   NaN  
50%         NaN    NaN  NaN        NaN          NaN   NaN   NaN  
75%         NaN    NaN  NaN        NaN          NaN   NaN   NaN  
max         NaN    NaN  NaN        NaN          NaN   NaN   NaN

4. In the example above, I use the dataset that contains a lot of missing data on `age` and `deck`. Here, if I use `df.dropna()` before `del df['col']` (in this case the `col` are `age` and `deck`), there are only 182 data left. However, if I use `del df['col']` before `df.dropna()`, there are still 889 data left (only 2 short from the original due to 2 missing values on the `embarked` column). So, if I think that a column has too much missing values, it's better to do `del df['col']` before `df.dropna()` rather than the reverse (or just `df.dropna()`).

## **8.**

1. `df.groupby("col1")["col2"].describe()` outputs a table that is fundamentally provides the same data as of what `df.describe()` would do (like mean, min, max, etc.), but instead of counting it for the whole dataset, it divides the dataset into groups first, according to the number of unique data in `col1` (for example, if there are 4 unique values, then there are 4 groups). Then, after that the code will provide the summarization of each group in the form of a table. Demonstration is shown below.

In [ ]:
df2.groupby("pclass")["age"].describe()

count       mean        std   min   25%   50%   75%   max
pclass                                                           
1       186.0  38.233441  14.802856  0.92  27.0  37.0  49.0  80.0
2       173.0  29.877630  14.001077  0.67  23.0  29.0  36.0  70.0
3       355.0  25.140620  12.495398  0.42  18.0  24.0  32.0  74.0

In [ ]:
df2.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

2. `count` variable in `df.groupby("col1")["col2"].describe()` only counts the number of non-missing data in the column `col2`, in respect to their `col1` grouping, while `df.describe()` counts number of non-missing data for each column in the dataset.
3. From all of the 7 errors, all of them are better to ask from ChatGPT rather than google, even though for some errors such as error A and E, google are "kinda" straightforward about it. Since ChatGPT has scanned the csv file before (at least the column names and the dimension of it) and knew about the `groupby` code, ChatGPT can quickly determine if this is an error because of a typo or other issues.

## **9.**

Uhh, I somewhat read some of it, although I might need to do some reviewing and google searching (or maybe even use ChatGPT to help me go through some confusion I have).

## ChatGPT link (for q6-q7): https://chatgpt.com/share/66e38a15-28d0-8012-bb79-6c48a4bc2546

## ChatGPT link (for q8): https://chatgpt.com/share/66e38a37-0e84-8012-969d-63e6a0ab9f31